libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,607 @@
+    // For GPS it isn't very accurate if we aren't moving...?
+
+    float direction = ODID_INV_DIR;
+    if (!got_bad_gps_fix) {
+        direction = wrap_360(degrees(ahrs.groundspeed_vector().angle())); // heading (degrees)
+    }
+
+    const float speed_horizontal = create_speed_horizontal(ahrs.groundspeed());
+
+    Vector3f velNED;
+    UNUSED_RESULT(ahrs.get_velocity_NED(velNED));
+    const float climb_rate = create_speed_vertical(-velNED.z); //make sure climb_rate is within Remote ID limit
+
+    int32_t latitude = 0;
+    int32_t longitude = 0;
+    if (current_location.check_latlng()) { //set location if they are valid
+        latitude = current_location.lat;
+        longitude = current_location.lng;
+    }
+
+    // FIXMME
+    // Q? The mavlink docs reference being calculated against  29.92 in HG. Presumably it is actually against the US Standard Atmosphere Standard Day
+    // need to be sure AP is outputting the correct barometric altitude frame here pressure and temperature assumptions.
+    float altitude_barometric = 0.0;
+    altitude_barometric = create_altitude(barometer.get_altitude());
+
+    // FIX OTHER GPS DRIVERS to output this too!!!
+
+    // Should we check if the GPS driver has ellipsoid height available???
+    float altitude_geodetic_cm = create_altitude(gps.height_above_ellipsoid() * 10.0);
+
+    // Compute the current height above the takeoff location
+    float height_above_takeoff = 0;     // height above takeoff (meters)
+    if (hal.util->get_soft_armed()) {
+        int32_t curr_alt_asml_cm;
+        int32_t takeoff_alt_asml_cm;
+        if (current_location.get_alt_cm(Location::AltFrame::ABSOLUTE, curr_alt_asml_cm) &&
+            _takeoff_location.get_alt_cm(Location::AltFrame::ABSOLUTE, takeoff_alt_asml_cm)) {
+            height_above_takeoff = (curr_alt_asml_cm - takeoff_alt_asml_cm) * 0.01;
+        }
+    }
+    float height = create_altitude(height_above_takeoff);
+
+    // Accuracy
+
+    // If we have GPS 3D lock we presume that the accuracies of the system will track the GPS's reported accuracy
+    MAV_ODID_HOR_ACC horizontal_accuracy_mav = MAV_ODID_HOR_ACC_UNKNOWN;
+    MAV_ODID_VER_ACC vertical_accuracy_mav = MAV_ODID_VER_ACC_UNKNOWN;
+    MAV_ODID_SPEED_ACC speed_accuracy_mav = MAV_ODID_SPEED_ACC_UNKNOWN;
+    MAV_ODID_TIME_ACC timestamp_accuracy_mav = MAV_ODID_TIME_ACC_UNKNOWN;
+
+    float horizontal_accuracy;
+    if (gps.horizontal_accuracy(horizontal_accuracy)) {
+        horizontal_accuracy_mav = create_enum_horizontal_accuracy(horizontal_accuracy);
+    }
+
+    float vertical_accuracy;
+    if (gps.vertical_accuracy(vertical_accuracy)) {
+        vertical_accuracy_mav = create_enum_vertical_accuracy(vertical_accuracy);
+    }
+
+    float speed_accuracy;
+    if (gps.speed_accuracy(speed_accuracy)) {
+        speed_accuracy_mav = create_enum_speed_accuracy(speed_accuracy);
+    }
+
+    timestamp_accuracy_mav =  create_enum_timestamp_accuracy(1.0); //set to 1 sec accuracy
```

Comment:
This needs to be set eventually.
As Peter said when we have GPS  it should report MAV_ODID_TIME_ACC_0_1_SECOND

But once we lose GPS how fast does the accuracy of the clock degrade will vary a lot per system but still likely less than 0.1 second for a long time after losing GPS

Reply:
comment added